tests/e2e/hashkv_test.go

```diff
@@ -0,0 +1,359 @@
+	secondCompactOnRev := firstCompactOnRev + 3
+	t.Logf("COMPACT rev=%d", secondCompactOnRev)
+	_, err = cli.Compact(ctx, secondCompactOnRev, config.CompactOption{Physical: true})
+	require.NoError(t, err)
+
+	for hashKVOnRev := secondCompactOnRev; hashKVOnRev <= dataset.lastestRevision; hashKVOnRev++ {
+		t.Logf("HashKV on rev=%d", hashKVOnRev)
+		resp, err := cli.HashKV(ctx, hashKVOnRev)
+		require.NoError(t, err)
+
+		require.Len(t, resp, 3)
+		require.True(t, resp[0].Hash != 0)
+		t.Logf("One Hash value is %d", resp[0].Hash)
+
+		require.Equal(t, resp[0].Hash, resp[1].Hash)
+		require.Equal(t, resp[1].Hash, resp[2].Hash)
+	}
+}
+
+// key: "foo"
+// modified: 14
+// generations:
+//
+//	{13, 0}, {14, 0}
+//	{2, 0},  {4, 0},  {6, 0}[tombstone]
+//
+// key: "foo1"
+// modified: 6
+// generations:
+//
+//	{empty}
+//	{3, 0}, {5, 0}, {6, 1}[tombstone]
+//
+// key: "foo2"
+// modified: 11
+// generations:
+//
+//	{empty}
+//	{7, 0}, {10, 0}, {11, 0}[tombstone]
+//
+// key: "foo3"
+// modified: 11
+// generations:
+//
+//	{empty}
+//	{8, 0}, {9, 0}, {12, 0}[tombstone]
+func newTestDataset2InCluster(t *testing.T, clus *fe2e.EtcdProcessCluster, cliCfg fe2e.ClientConfig) *datasetInfo {
+	t.Helper()
+	c := newClient(t, clus.EndpointsGRPC(), cliCfg)
+	defer c.Close()
+
+	ctx := context.Background()
+
+	info := &datasetInfo{
+		keys: map[string]struct{}{
+			"foo":  {},
+			"foo1": {},
+			"foo2": {},
+			"foo3": {},
+		},
+	}
+
+	_, err := c.Put(ctx, "foo", "2")
+	require.NoError(t, err)
+
+	_, err = c.Put(ctx, "foo1", "3")
+	require.NoError(t, err)
+
+	_, err = c.Put(ctx, "foo", "4")
+	require.NoError(t, err)
+
+	_, err = c.Put(ctx, "foo1", "5")
+	require.NoError(t, err)
+
+	tresp, err := c.Txn(ctx).Then(
+		clientv3.OpDelete("foo"),
+		clientv3.OpDelete("foo1"),
+	).Commit()
+	require.NoError(t, err)
+	info.tombstones = append(info.tombstones, tresp.Header.Revision)
+
+	_, err = c.Put(ctx, "foo2", "7")
+	require.NoError(t, err)
+
+	_, err = c.Put(ctx, "foo3", "8")
+	require.NoError(t, err)
+
+	_, err = c.Put(ctx, "foo3", "9")
+	require.NoError(t, err)
+
+	_, err = c.Put(ctx, "foo2", "10")
+	require.NoError(t, err)
+
+	dresp, err := c.Delete(ctx, "foo2")
+	require.NoError(t, err)
+	info.tombstones = append(info.tombstones, dresp.Header.Revision)
+
+	dresp, err = c.Delete(ctx, "foo3")
+	require.NoError(t, err)
+	info.tombstones = append(info.tombstones, dresp.Header.Revision)
+
+	_, err = c.Put(ctx, "foo", "13")
+	require.NoError(t, err)
+
+	resp, err := c.Put(ctx, "foo", "14")
+	require.NoError(t, err)
+	require.Equal(t, int64(14), resp.Header.Revision)
+	info.lastestRevision = resp.Header.Revision
+	return info
+}
+
+func newClusterForHashKV(t *testing.T, clusVersion fe2e.ClusterVersion) (*fe2e.EtcdProcessCluster, fe2e.ClientConfig) {
+	if clusVersion != fe2e.CurrentVersion {
+		if !fileutil.Exist(fe2e.BinPath.EtcdLastRelease) {
+			t.Skipf("%q does not exist", fe2e.BinPath.EtcdLastRelease)
+		}
+	}
+
+	ctx := context.Background()
+
+	fe2e.BeforeTest(t)
```

Comment:
These should be in the test case directly instead of a helper function.

I don't think we need `newClusterForHashKV` at all.

Reply:
OK. I removed it.